### PR TITLE
persist: allow specifying optional role ARN for S3 usage

### PR DESCRIPTION
**Pull request by danhhz**
### Motivation

  * This PR adds a feature that has not yet been specified.

Role ARN support for cloud.

### Tips for reviewer

Verified manually by creating an s3 bucket in my personal AWS account with the following bucket policy that give access to that bucket's `foo/*` path to a role I can assume in the mz scratch AWS account.

    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Principal": {
                    "AWS": "arn:aws:iam::REDACTED:role/engineer"
                },
                "Action": "s3:*",
                "Resource": "arn:aws:s3:::REDACTED/foo/*"
            },
            {
                "Effect": "Allow",
                "Principal": {
                    "AWS": "arn:aws:iam::REDACTED:role/engineer"
                },
                "Action": "s3:ListBucket*",
                "Resource": "arn:aws:s3:::REDACTED",
                "Condition": {
                    "StringLike": {
                        "s3:prefix": "foo/*"
                    }
                }
            }
        ]
    }

Authing with the scratch account without the role doesn't allow access.

    $ cargo run -- --dev --experimental --persist-storage=s3://REDACTED/foo
    materialized: persistence error: Error { code: "AccessDenied", message: "Access Denied", request_id: "27DW57MQZKZEPHAG", s3_extended_request_id: "uEoMFoqG7p7CBwVQofjT9lMHqSXYtZ5O6/qE/NiuOVDfZTSvc6gU2gRfOZPsPs+Nz4QpY78uPBk=" }

Adding the role works!

    $ cargo run -- --dev --experimental --persist-storage=s3://REDACTED/foo?aws_role_arn=arn%3Aaws%3Aiam%3A%3AREDACTED%3Arole%2Fengineer
    =======================================================================
    Thank you for trying Materialize!
    ...

And persist data starts appearing in the bucket.

Pointing at a different path in the bucket also doesn't allow access.

    $ cargo run -- --dev --experimental --persist-storage=s3://REDACTED/foo?aws_role_arn=arn%3Aaws%3Aiam%3A%3AREDACTED%3Arole%2Fengineer
    materialized: persistence error: Error { code: "AccessDenied", message: "Access Denied", request_id: "HZF9CETHWGZB4JK5", s3_extended_request_id: "RSa09MwzDSGLd02/GY5vhi7dp9cT0tRf1R9p7Ud4k1r4JXoMy+JYrINNErMnuD3fKyPDzU0qQxU=" }

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/persist/src/s3.rs [hunk 1]
```diff
@@ -49,8 +51,22 @@ impl Config {
...
-    pub async fn new(bucket: String, prefix: String) -> Result<Self, Error> {
-        let config = AwsConfig::load_from_env().await;
+    pub async fn new(
+        bucket: String,
+        prefix: String,
+        role_arn: Option<String>,
+    ) -> Result<Self, Error> {
+        let mut config = AwsConfig::load_from_env().await;
+        if let Some(role_arn) = role_arn {
+            let provider = AssumeRoleProvider::builder(role_arn).session_name("persist");
```
Comment (danhhz) [inline src/persist/src/s3.rs:61]: TODO: figure out session_name and external_id

File: src/coord/src/persistcfg.rs [hunk 2]
```diff
@@ -58,7 +59,13 @@ impl TryFrom<String> for PersistStorage {
                     .strip_prefix('/')
                     .unwrap_or_else(|| url.path())
                     .to_string();
-                Ok(PersistStorage::S3(PersistS3Storage { bucket, prefix }))
+                let mut query_params = url.query_pairs().collect::<HashMap<_, _>>();
+                let role_arn = query_params.remove("role_arn").map(|x| x.into_owned());
```
Comment (danhhz) [inline src/coord/src/persistcfg.rs:63]: Our snowflake precedent starts to fall apart here a bit, they have these as SQL params: https://docs.snowflake.com/en/sql-reference/sql/create-storage-integration.html#cloud-provider-parameters-cloudproviderparams

Also, I should probably call this "aws_role_arn" or "storage_aws_role_arn" to match.

**Comment by danhhz** [comment]
Bonus note to self: Alex requested that whatever we do here is backward compatible with old versions of mz. However, as currently written, the versions of mz that have this param will happily parse the new format and throw away the role_arn, switching the background tables test to use s3, but failing with permissions errors because we didn't set the role. Relatedly, we probably also want a cli flag to opt the background system tables test into this, so we can incrementally roll it out: hook things up, test it manually for a while with the cli flag override, then finally flip it on for some customers.

**Comment by danhhz** [comment]
> Alex requested that whatever we do here is backward compatible with old versions of mz.

Hmm, can we even do that? Even if we made a new flag, old mz versions would error out that they don't understand it: 
```
$ materialized --foo
error: Found argument '--foo' which wasn't expected, or isn't valid in this context

USAGE:
    materialized [OPTION]...

For more information try --help
```

**Comment by danhhz** [comment]
cc @alex-hunt-materialize

**Comment by alex-hunt-materialize** [comment]
This is a point in favor of ENV vars over command line args, since they can be safely ignored by versions that don't understand them. Another option is to not begin passing the command line arg in Cloud until all customers are upgraded to a version that supports it.

**Review by benesch — APPROVED** [review]
This LGTM! It seems to me that it's probably easiest for cloud to learn to change the set of CLI flags it boots the pod with based on the Materialize version. The action-at-a-distance of env vars stresses me out. I don't love the idea of cloud setting an env var that it expects to be ignored, but only by some versions of Materialize.

**Comment by danhhz** [comment]
Managed to manually verify this! Put instructions in the PR description

**Comment by danhhz** [comment]
Also added a `--persist-storage-enabled` feature gate that makes `--persist-storage` a no-op if unset. `--persist-storage` no longer requires experimental, but `--persist-storage-enabled` does.

Combined, this means it should be safe for mz cloud to start specifying `--persist-storage` as of the mz release that contains this PR. If it's specified to a previous version, it'll error out with either a "you need --experimental" or a "unknown flag" error.

Once they do, trying out s3 based persist with the background system tables test will be as simple as: `--persist-storage-enabled --experimental`

**Pull request merged**